### PR TITLE
set vm ancestry from relationship resource info

### PR DESCRIPTION
**Pull request by d-m-u**
we are working on converting vm genealogy to use ancestry. In order to do this, we need to use the existing relationship resource information on all vms with genealogy rels of type VmOrTemplate to convert those resource ids into vm ids. 

This PR 1) adds the necessary index and 2) gets the ancestry information from the relationships, splitting on slashes, then maps those back to the vm ancestry records. 

the first commit's the original sql which joined to vms
the second commit removes that join so we no longer have a potential issue with the ordering 

needs https://github.com/ManageIQ/manageiq/pull/20788, tests are green: https://github.com/ManageIQ/manageiq-cross_repo-tests/pull/225

### CODE REVIEW

**Review by Fryguy — DISMISSED** [review]
This is awesome.  Great idea to get genealogy out of there!

**Comment by d-m-u** [comment]
hey @Fryguy could I please ask you to review this again?

**Comment by d-m-u** [comment]
could I please get someone to review this?

**Comment by d-m-u** [comment]
commenting so I don't forget: it's work in progress because I think Keenan's right about what's missing: "... is [it] possible that not all the ids are in the database - could that be the issue?"

**Comment by kbrock** [comment]
aside: we have some notes in [gist](https://gist.github.com/kbrock/3ee8e721f0e50218621652ac55b81652) - may want to bring that into this PR for future reference (or not...)

**Comment by NickLaMuro** [comment]
Ran this migration against a large customer database and it passed without issue.

Does that mean I understand what this SQL is doing?  Nope!

But it didn't blow up.  🎉

**Comment by d-m-u** [comment]
@kbrock I'm pretty convinced that the order (and everything else) on this is correct and at this point I'm not sure how else to test that.

**Comment by d-m-u** [comment]
hey @agrare could I please ask you to review this? I think you might have a stake here.

**Comment by d-m-u** [comment]
Okay, well. Even though we consensus'd on closing [the pr from hell](https://github.com/ManageIQ/manageiq/pull/20274) because it has too much overhead to realistically be a viable next step, I still think this PR was both an excellent idea and low enough overhead to make it worth merging. I don't know what else I need to do to get anyone else to believe me about that (thanks for the review, @NickLaMuro).

**Comment by kbrock** [comment]
outstanding:

we need to fix the "PR from hell"
To do that, we need to remove relationships from vms
To do that, we need to introduce relationships to `ems_folders`

There is a possible issue around loading resource_pools - but I think this will resolve itself.

**Comment by d-m-u** [comment]
thanks @kbrock

**Comment by Fryguy** [comment]
Ignore the rubocops...I've added them to the ignore list here - https://github.com/ManageIQ/manageiq-style/pull/8

File: spec/migrations/20200607025146_add_ancestry_to_vm_spec.rb
Comment (kbrock) [reply spec/migrations/20200607025146_add_ancestry_to_vm_spec.rb:38]: They say comments are a code smell.
The same folks also say to comment your code.
Sigh.

So the only way to do that is to create self documenting code.

https://github.com/ManageIQ/manageiq/blob/master/spec/models/mixins/relationship_mixin_spec.rb#L3-L7

Sorry, just remembered that code when I was typing the last response. It was amusing that the code was for relationship mixin, since we are ... yea

If you look at the evolution from your first version to the second version, and you extrapulate, you'll see how someone ended up here.

File: db/migrate/20200607025146_add_ancestry_to_vm.rb
Comment (kbrock) [inline db/migrate/20200607025146_add_ancestry_to_vm.rb:26]: I know we had to put in ruby code that mimics the sql because it is just too much.

but (at least now when the code is fresh in my mind) like the end result.

getting this into a migration routine may be nice - not sure if we want something like this in ancestry.
Definitely want a better way to populate ancestry based upon sql. (sorry tangent)
^ ignore the ancestry comment. want to keep the thought out there but  don't want us to move towards that right now

**Review by kbrock — COMMENTED** [review]
sorry - I know we went over the specs a few times. this link was itching at the back of my mind and only surfaced today

File: db/migrate/20200607025146_add_ancestry_to_vm.rb
Comment (d-m-u) [reply db/migrate/20200607025146_add_ancestry_to_vm.rb:26]: `but (at least now when the code is fresh in my mind) like the end result.`
Are you missing a verb? Is this saying that you do like the end result here now, or that you do not?

Comment (kbrock) [reply db/migrate/20200607025146_add_ancestry_to_vm.rb:26]: sorry, missed a noun. I **do** like the end result

**Comment by kbrock** [comment]
this looks nice.

I was thinking that useing https://github.com/ManageIQ/manageiq/blob/master/spec/models/mixins/relationship_mixin_spec.rb#L870-L886 would allow you to pass in a hash and not need to create any of the objects.

If this is more clear for you then this is good

**Comment by d-m-u** [comment]
@kbrock sorry, yes. the reason i chose to not go with that suggestion is that that particular code you linked to relies on 1) the parent setter which relies on 2) with_relationship_type which relies on 3) caching methods that I didn't really want to bring into all this. My choice was either to rewrite that and keep the functionality without the extra code, or this approach, which felt ... easier.

File: db/migrate/20200607025146_add_ancestry_to_vm.rb
Comment (kbrock) [reply db/migrate/20200607025146_add_ancestry_to_vm.rb:26]: um, I missed the "I" which is a pro noun.

I guess I then added the do to answer your question of do or do not (that is the question).

So I stand by my noun statement.

**Comment by kbrock** [comment]
hey @d-m-u - where did we leave this? were the spec changes good to go?
I figure we're going to duplicate those tests across a few migrations so good to have a pattern together.

Also, did you decide to pull the migration out into a separate class?

**Comment by d-m-u** [comment]
marking as work in progress: i'm rewriting the specs per the last comment

**Review by kbrock — COMMENTED** [review]
sorry - could you pull out the ancestry_resource_ids into an AncestryMigrationhelper

then we can ship this

**Comment by kbrock** [comment]
sigh - sorry that I had you change perfectly working code

File: db/migrate/20200607025146_add_ancestry_to_vm.rb
Comment (kbrock) [inline db/migrate/20200607025146_add_ancestry_to_vm.rb:2]: this should be loaded via [spec_helper]
Probably want to scope this

```suggestion
  include ::RelationshipAncestryConverter
```

[spec_helper]: https://github.com/ManageIQ/manageiq-schema/blob/d30ccc15125a7c19b23182ba46bd96a05a12e9fb/spec/spec_helper.rb#L14

if that doesn't work, wonder if using extend `instead` would do what you need - but 
include` seems to be the right thing to me right now

**Review by kbrock — COMMENTED** [review]
thanks for agreeing to refactor code that was working

File: db/migrate/20200607025146_add_ancestry_to_vm.rb
Comment (d-m-u) [reply db/migrate/20200607025146_add_ancestry_to_vm.rb]: thank you  👍

**Comment by d-m-u** [comment]
marking [WIP] because I'm mildly concerned about the test failure (on https://github.com/ManageIQ/manageiq/pull/20788): we can't serialize objects with unbound methods, so anything previously that got queued that has our ancestry binding on latest code will be broken.

**Comment by d-m-u** [comment]
https://github.com/ManageIQ/manageiq-cross_repo-tests/pull/225
yay

File: lib/relationship_ancestry_converter_helper.rb
Comment (Fryguy) [inline lib/relationship_ancestry_converter_helper.rb:1]: Are you planning on using this for other tables in future migrations?  If not, might just be better to embed this module directly into the migration that uses it.

Comment (d-m-u) [reply lib/relationship_ancestry_converter_helper.rb:1]: Yes. We're planning on using this for https://github.com/ManageIQ/manageiq-schema/pull/497 as well.

Comment (Fryguy) [reply lib/relationship_ancestry_converter_helper.rb:1]: Is it possible to write specs for it as a standalone module or is that kind of impossible?

File: lib/relationship_ancestry_converter_helper.rb
Comment (Fryguy) [inline lib/relationship_ancestry_converter_helper.rb:41]: I'm curious about these commented bits...are they for the other PR in the future?  If so, should they just be added at that time?

Comment (d-m-u) [reply lib/relationship_ancestry_converter_helper.rb:41]: I wanted ruby versions of the sql for each method just to make it easier to understand.

Comment (Fryguy) [reply lib/relationship_ancestry_converter_helper.rb:41]: ohhhh I see - I didn't realize they were the same/related method but Ruby - wonder if it's better to organize them together - Something like:

Before:

```ruby
  # def ancestry_resource_ids(relationship, resource_type, id_range)
  #   Relationship.where(
  #     :relationship => relationship,
  #     :resource_type => resource_type,
  #     :resource_id => id_range
  #   ).where.not(:ancestry => nil).flat_map do |a_rels|
  #     a_rels.ancestry.split('/').each_with_index.map { |rel_id, indx| [a_rels.resource_id, rel_id, indx] }
  #   end
  # end

  def ancestry_resource_ids(relationship, resource_type, id_range)
    <<-SQL
    SELECT a_rels.resource_id AS src_id, relationships.id AS rel_id, relationships.indx AS rel_indx
    FROM   relationships a_rels
    LEFT JOIN LATERAL UNNEST(STRING_TO_ARRAY(a_rels.ancestry, '/')::BIGINT[])
    WITH ORDINALITY AS relationships(id, indx) ON TRUE
    WHERE  a_rels.relationship = '#{relationship}'
    AND    a_rels.resource_type = '#{resource_type}'
    AND    a_rels.resource_id BETWEEN #{id_range.first} AND #{id_range.last}
    SQL
  end
```

After:

```ruby
  def ancestry_resource_ids(relationship, resource_type, id_range)
    <<-SQL
    SELECT a_rels.resource_id AS src_id, relationships.id AS rel_id, relationships.indx AS rel_indx
    FROM   relationships a_rels
    LEFT JOIN LATERAL UNNEST(STRING_TO_ARRAY(a_rels.ancestry, '/')::BIGINT[])
    WITH ORDINALITY AS relationships(id, indx) ON TRUE
    WHERE  a_rels.relationship = '#{relationship}'
    AND    a_rels.resource_type = '#{resource_type}'
    AND    a_rels.resource_id BETWEEN #{id_range.first} AND #{id_range.last}
    SQL

    # The above is a pure-SQL version of this similar Ruby code:
    #
    # Relationship.where(
    #   :relationship => relationship,
    #   :resource_type => resource_type,
    #   :resource_id => id_range
    # ).where.not(:ancestry => nil).flat_map do |a_rels|
    #   a_rels.ancestry.split('/').each_with_index.map { |rel_id, indx| [a_rels.resource_id, rel_id, indx] }
    # end
  end
```

File: lib/relationship_ancestry_converter_helper.rb
Comment (chessbyte) [inline lib/relationship_ancestry_converter_helper.rb:46]: redundant method definition in the comment.  In other words, suggest to remove the `def ancestry_resource_ids(...)` in the comments and the correlating `end`

Comment (d-m-u) [reply lib/relationship_ancestry_converter_helper.rb:46]: I don't agree. I think the definitions are useful here because tracking the method params has been really important for catching bugs and understanding what this is doing.

File: lib/relationship_ancestry_converter_helper.rb
Comment (d-m-u) [reply lib/relationship_ancestry_converter_helper.rb]: Even with the redundant method signatures, @Fryguy, you struggled to understand what this is doing, so I don't exactly think taking them out is going to help. This PR's kind of fragile, and this is about the third attempt of this PR.

Comment (Fryguy) [reply lib/relationship_ancestry_converter_helper.rb]: No I think combining them together makes the SQL easier to understand, which was my biggest struggle...I can use the two together to understand the intent.

File: lib/relationship_ancestry_converter_helper.rb
Comment (kbrock) [reply lib/relationship_ancestry_converter_helper.rb:1]: it is going to be used 3 times max? I'd honestly prefer copy/past over spending the money/time on testing this as an individual component. The output of the code is already tested for this particular use case. and will be tested for the other use case as well.

**Review by kbrock — APPROVED** [review]
Feel like this has dragged a bit and I have already asked to rewrite the sql a few times and then extract it.

Lets get this to a reasonable point and get it in

**Comment by Fryguy** [comment]
We can't merge it until the other side is available (or is it done? don't see any linked PRs in the OP, and there are a lot of comments here, so not sure)

**Comment by d-m-u** [comment]
https://github.com/ManageIQ/manageiq/pull/20788

**Comment by d-m-u** [comment]
Could I get this merged, please?

**Comment by kbrock** [comment]
@Fryguy I too would like to bump this.

Premise:

```ruby
add_column :vms, :ancestry, :string
Vm.update(:ancestry => vms.relationships.ancestry)
Vm.relationships.delete_all
```

Of course the devil in the details has a cascading effect:

1. Change ancestry from `relationships.id` to `vms.id`.
2. Change `O(N + log(N))` to `O(1)` since the substitution thing mentioned above is is more a tad complex than it looks.
3. The sql needed for `O(1)` is more obtuse than the ruby version.

**Comment by d-m-u** [comment]
Okay, I'm marking this and the core pr (https://github.com/ManageIQ/manageiq/pull/20788) as work in progress again because I'd like to rewrite the tests.

**Comment by d-m-u** [comment]
@kbrock could I get another look at this please?

**Review by kbrock — APPROVED** [review]
this looks nice.

ship it

**Review by Fryguy — APPROVED** [review]
I think this is good, but would like @agrare to also review and approve.

File: spec/migrations/20200607025146_add_ancestry_to_vm_spec.rb
Comment (Fryguy) [inline spec/migrations/20200607025146_add_ancestry_to_vm_spec.rb:171]: Minor, but should this raise an Exception instead?  Either way, this is a debugging statement so it should probably be removed as per the bot's request.

File: db/migrate/20200607025146_add_ancestry_to_vm.rb
Comment (Fryguy) [inline db/migrate/20200607025146_add_ancestry_to_vm.rb:41]: SUPER minor, but 
```suggestion
    children = src_model.select("ancestry::bigint").where("ancestry NOT LIKE '%/%'")
```

**Comment by Fryguy** [comment]
From a testing perspective, have you run this against a "real" customer database?  I'm curious if it migrates correctly.  If so, that would give a huge confidence boost.  If it fails, then that's more test cases.

**Comment by d-m-u** [comment]
@Fryguy yes, @NickLaMuro was awesome and provided the customer database that was used to test this. Please see his earlier comment: https://github.com/ManageIQ/manageiq-schema/pull/492#issuecomment-659679138

**Comment by d-m-u** [comment]
@agrare sorry yes, it's schema then core then the providers

**Review by agrare — COMMENTED** [review]
:+1: LGTM after @Fryguy 's two suggestions are fixed

**Pull request merged**
